lib/url.js

```diff
@@ -683,40 +495,575 @@ Url.prototype.resolveObject = function(relative) {
+      this._parseHash(str, i, end);
+      pathEnd = i - 1;
+      break;
+    } else if (ch === 0x3F /*'?'*/) {
+      this._parseQuery(str, i, end);
+      pathEnd = i - 1;
+      break;
+    } else if (!escape && autoEscapeCharacters[ch] === 1) {
+      escape = true;
+    }
+  }
+
+  if (pathStart > pathEnd) {
+    this.pathname = prePath === '' ? '/' : prePath;
+    return;
+  }
+
+  var path;
+  if (escape) {
+    path = this._getComponentEscaped(str, pathStart, pathEnd, false);
+  } else {
+    path = str.slice(pathStart, pathEnd + 1);
+  }
+  this.pathname = prePath === '' ?
+      (this._prependSlash ? '/' + path : path) : prePath + path;
+};
+
+Url.prototype._parseQuery = function(str, start, end) {
+  var queryStart = start;
+  var queryEnd = end;
+  var escape = false;
+  var autoEscapeCharacters = _autoEscapeCharacters;
+
+  for (var i = start; i <= end; ++i) {
+    var ch = str.charCodeAt(i);
+
+    if (ch === 0x23 /*'#'*/) {
+      this._parseHash(str, i, end);
+      queryEnd = i - 1;
+      break;
+    } else if (!escape && autoEscapeCharacters[ch] === 1)
+      escape = true;
+  }
+
+  if (queryStart > queryEnd) {
+    this.search = '';
+    return;
+  }
+
+  var query = escape ?
+      this._getComponentEscaped(str, queryStart, queryEnd, true) :
+      str.slice(queryStart, queryEnd + 1);
+
+  this.search = query;
+};
+
+Url.prototype._parseHash = function(str, start, end) {
+  if (start > end) {
+    this.hash = '';
+    return;
+  }
+  this.hash = this._getComponentEscaped(str, start, end, true);
+};
+
+Object.defineProperty(Url.prototype, 'port', {
+  get: function() {
+    if (this._port >= 0)
+      return ('' + this._port);
+    return null;
+  },
+  set: function(v) {
+    if (v == null)
+      this._port = -1;
+    else
+      this._port = parseInt(v, 10);
+  },
+  enumerable: true,
+  configurable: true
+});
+
+Object.defineProperty(Url.prototype, 'query', {
+  get: function() {
+    var query = this._query;
+    if (query != null)
+      return query;
+
+    var search = this.search;
+
+    if (search) {
+      if (search.charCodeAt(0) === 0x3F /*'?'*/)
+        search = search.slice(1);
+      if (search !== '') {
+        this._query = search;
+        return search;
+      }
+    }
+    return search;
+  },
+  set: function(v) {
+    if (typeof v === 'string') {
+      if (v !== '') this.search = '?' + v;
+      this._query = v;
+    } else if (v !== null && typeof v === 'object') {
+      var string = querystring.stringify(v);
+      this.search = string !== '' ? '?' + querystring.stringify(v) : '';
+      this._query = v;
+    } else {
+      this._query = this.search = null;
+    }
+  },
+  enumerable: true,
+  configurable: true
+});
+
+Object.defineProperty(Url.prototype, 'path', {
+  get: function() {
+    var p = this.pathname || '';
+    var s = this.search || '';
+    if (p || s)
+      return p + s;
+    return (p == null && s) ? ('/' + s) : null;
+  },
+  set: function(path) {
```

Comment:
we might check for false-y values like `null` or `undefined` here and short-circuit (also on `host`).